enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+`DNSNameResolver` CR.
+* On receiving the delete event, the `DNSNameResolver` controller will remove the details stored regarding the `DNSNameResolver` CRs.
+* If the deleted event is for a `DNSNameResolver` corresponding to a regular DNS name, the `DNSNameResolver` controller will remove the details of a regular DNS
+name, if it doesn't match a `DNSNameResolver` CR corresponding to a  wildcard DNS name.
+* If the deleted event is for a `DNSNameResolver` corresponding to a wildcard DNS name, the `DNSNameResolver` controller will remove the details of a wildcard DNS
+name. It will also remove the details of the regular DNS names added to the `.status` of the CR, if each of them doesn't match a `DNSNameResolver` CR corresponding
+to a  wildcard DNS name.
+
+#### Workflow of `ocp_dnsnameresolver` CoreDNS plugin
+
+The following pseudocodes explain the workflow of the `ocp_dnsnameresolver` CoreDNS plugin:
+
+##### Watch
+```shell
+watch(DNSNameResolver)
+
+if match(event.type, "create"):
+
+	if is_wildcard(obj.spec.name) == true:
+		add_wildcard_dns_name(obj.spec.name)
+	else
+		add_regular_dns_name(obj.spec.name)
+
+else if match(event.type, "delete"):
+
+	if is_wildcard(obj.spec.name) == true:
+		delete_wildcard_dns_name(obj.spec.name)
+	else
+		delete_regular_dns_name(obj.spec.name)
+```
+
+* The `ocp_dnsnameresolver` CoreDNS plugin will watch for the events related to `DNSNameResolver` CRs and will store the DNS name along
+with the corresponding CR name.
+* On receiving the delete event, the `ocp_dnsnameresolver` plugin will remove the details stored regarding the `DNSNameResolver` CRs.
+
+##### DNS lookup interception
+
+```shell
+response = get_response_from_plugin_chain(dnsName)
+
+if match_regular_dns_name(dnsName) == false || match_wildcard_dns_name(dnsName) == false:
+	return response
+
+if resolution_failed(response) == false:
+
+	ips, ttls = get_ips_and_ttls(response)
+	for each ttl in ttls {
+		if ttl = 0:
+			ttl = min_ttl
+	}
+	
+	if match_regular_dns_name(dnsName) == true:
+		objName = get_regular_dns_name_resolver_name(dnsName)
+		update_dns_resolver_success(objName, dnsName, ips, ttls, response)
+	
+	if match_wildcard_dns_name(dnsName) == true:
+		objName = get_wildcard_dns_name_resolver_name(dnsName)
+		update_dns_resolver_success(objName, dnsName, ips, ttls, response)
+
+else
+  	
+	if match_regular_dns_name(dnsName) == true:
+		objName = get_regular_dns_name_resolver_name(dnsName)
+		update_dns_resolver_failure(objName, dnsName, response)
+	
+	if match_wildcard_dns_name(dnsName) == true:
+		objName = get_wildcard_dns_name_resolver_name(dnsName)
+		update_dns_resolver_failure(objName, dnsName, response)
+
+return response
+```
+
+* The `ocp_dnsnameresolver` CoreDNS plugin will intercept the request and the response for the DNS lookups from all the pods. If the DNS name matches
```

Comment:
Does the plugin need to intercept both the request and response?  Why not just the response?  What if it updates the ttl in the corresponding `.status` for a request but it gets no response?  Shouldn't the ttl update be based on the receipt of the response and not the receipt of the request?

Reply:
The CoreDNS plugin chain works in this way only. The request will be received first by the plugin then the response. When the request is received no actions are performed. It is sent to the next plugin in the chain. Actions are performed only after receiving the response. To be explicit I mentioned that it will intercept the request and the response of the the DNS lookups.